protos/flyteidl/plugins/spark.proto

```diff
@@ -31,4 +33,24 @@ message SparkJob {
     // Domain name of your deployment. Use the form <account>.cloud.databricks.com.
     // This instance name can be set in either flytepropeller or flytekit.
     string databricksInstance = 9;
+    // The pod spec and metadata to be used as the base configuration when creating the driver Pod for this task.
+    // +optional
+    oneof driverPodValue {
+        core.K8sPod driverPod = 10;
+    }
+    // Reference to an existing PodTemplate k8s resource to be used as the base configuration when creating the 
+    // driver Pod for this task. If this value is set, the specified PodTemplate will be used instead of, but applied
+    // identically as, the default PodTemplate configured in FlytePropeller.
+    // +optional
+    string driverPodTemplateName = 11;
+    // The pod spec and metadata to be used as the base configuration when creating the executor Pods for this task.
+    // +optional
+    oneof executorPodValue {
+        core.K8sPod executorPod = 12;
+    }
+    // Reference to an existing PodTemplate k8s resource to be used as the base configuration when creating the 
+    // executor Pods for this task. If this value is set, the specified PodTemplate will be used instead of, but applied
+    // identically as, the default PodTemplate configured in FlytePropeller.
+    // +optional
+    string executorPodTemplateName = 13;
```

Comment:
Is there any reason these can't just be the PodTemplate / PodTemplateName from the TaskExecutionContext? IIRC this is how we built some of the other plugins (ie. Dask, KF operator).

Reply:
Do you mean use the same pair for both executor and driver, use the existing pair from TaskExecutionContext for one of these and plumb the other here, or something else?

Reply:
Either or. Both of these are available in the task decorator in flytekit, so if we added them both at the `spark_config` level -- ex:
```
@task(
    task_config=Spark(
        spark_config=[
            executor_pod_template=...
            driver_pod_template=...
        ]
    )
    pod_template=odTemplate(
        primary_container_name="primary",
        labels={"lKeyA": "lValA", "lKeyB": "lValB"},
        annotations={"aKeyA": "aValA", "aKeyB": "aValB"},
        pod_spec=V1PodSpec(
            volumes=[V1Volume(name="volume")],
            tolerations=[
                V1Toleration(
                    key="num-gpus",
                    operator="Equal",
                    value=1,
                    effect="NoSchedule",
                ),
            ],
        )
)
```
it doesn't make sense to me intuitively how they would be applied. For things like resources / etc that are already available in the task decorator we implemented (in other plugins) that if set at the task-level they would be applied to all replicas (ie. executor / driver) unless there was a specific override there. So IMO we should apply task-level to the executor / driver by default and put an override on the driver only.

Reply:
I waffled on this a bit since the spark on k8s op itself has moved to an abstraction of explicit configs for each driver and executor, but yeah I tend to agree here that utilizing the top level `TaskTemplate.PodTemplate` as the primary override for both + a driver-specific `PodTemplate` in the SparkJob is more consistent with flytekit and other plugins. 

We'll need to clarify the order of precedence w.r.t. top level configs (i.e., TaskTemplate.Container), top level pod template (i.e., TaskTemplate.PodTemplate.PrimaryContainerName), and driver-specifc config (i.e., SparkJob.PodTemplate.PrimaryContainerName). The same set of clarifications are needed for other configs (i.e., memory/cores, which can also be specified as spark config -- `spark.executor.requests.memory`).

So the proposed python devx would be
```
@task(
    task_config=Spark(
        spark_conf={
           "spark.driver.memory": "4G",
           "spark.executor.memory": "8G",
        },
        # driver only pod template overrides (takes precedence over top level pod_template)
        driver_pod_template=...
    ),
    # default pod template overrides for BOTH driver and executor
    pod_template=odTemplate(
        primary_container_name="primary",
        labels={"lKeyA": "lValA", "lKeyB": "lValB"},
        annotations={"aKeyA": "aValA", "aKeyB": "aValB"},
        pod_spec=V1PodSpec(
            volumes=[V1Volume(name="volume")],
            tolerations=[
                V1Toleration(
                    key="num-gpus",
                    operator="Equal",
                    value=1,
                    effect="NoSchedule",
                ),
            ],
        )
    ),
)
```

@pingsutw @wild-endeavor any thoughts?

Reply:
I agree it's very confusing to have `podTemplate` in both `@task` and `spark` config, but I still think we need `worker_pod_template` in spark config. 

People may only want to override the `worker_pod_template`, for example, add gpu selector, or something else.

Reply:
i'd personally rather have the original where both pod templates are in the spark config, and then just error if the pod level config is specified.  but i'm okay any way.  I just feel like applying one could make equally convincing arguments that the normal top level config should be for executors or for drivers.

Reply:
Thanks for the thoughtful feedback. After weighing this + the suggestions in the [migrated PR](https://github.com/flyteorg/flyte/pull/4179#pullrequestreview-1662713211), I'd like to propose
* A RoleSpec definition in common.proto
* Inclusion of driver and executor role specs in SparkJob
* Error in spark plugin if TaskTemplate.PodTemplate is set
* Use this pattern as a deprecation path for TaskTemplate.PodTemplate for Dask, MPI, PyTorch, and Ray should we choose to have per-role PodTemplates/overrides (i.e., start erroring if both TaskTemplate.PodTemplate and worker RoleSpec are specified

Will summarize fully in https://github.com/flyteorg/flyte/pull/4179 and try to move the remainder of the communication there